Pscan rule tweaks

psiinon opened the pull request:
As per https://groups.google.com/d/msg/zaproxy-develop/6IAnEHANWpI/zva5350aBgAJ

kingthorin commented on src/org/zaproxy/zap/extension/pscanrules/CrossDomainScriptInclusionScanner.java:
```diff
@@ -60,14 +61,24 @@ public void scanHttpResponseReceive(HttpMessage msg, int id, Source source) {
+								 * 1. pscan rules cant make new requests and
+								 * 2. the browser will check it anyway
+								 */
+								continue;
```
Might be more clearn if the logic is reversed here and the alert raising moved in. Or if you really like this the comment should make it clear that the continue moves to the next sourceElement because integrity was found on the current one.

kingthorin commented on src/org/zaproxy/zap/extension/pscanrules/XFrameOptionScanner.java:
```diff
@@ -156,7 +159,8 @@ private String getMetaXFOEvidence(Source source) {
 				httpEquiv = metaElement.getAttributeValue("http-equiv");
 				if (HttpHeader.X_FRAME_OPTION.equalsIgnoreCase(httpEquiv)) {
-					return httpEquiv;
+					//return httpEquiv;
```
Could be removed?

kingthorin commented on src/org/zaproxy/zap/extension/pscanrules/resources/help/contents/pscanrules.html:
```diff
@@ -29,13 +29,18 @@
+By looking at the "src" attributes of "script" tags in the response.<br>
+Allowed Cross-Domain scripts:  
+<ul>
+<li>Any script with a non empty "integrity" attribute is ignored - the integrity value is not checked as should will be checked by the browser</li>
```
I think "non-empty" Should be hyphenated. Also something wrong toward the end "as should will"...

thc202 replied:
The last part of the sentence does not sound right ("... as it will be checked..." instead?).

kingthorin commented on src/org/zaproxy/zap/extension/pscanrules/resources/help/contents/pscanrules.html:
```diff
@@ -60,9 +65,10 @@
-types. The following conditions may result in an alert:
+This scanner checks for the existence and validity of the X-Frame-Options header.<br> 
+At MEDIUM and HIGH thresholds this only looks at non error HTML responses.<br>
+At LOW threshold it looks at all test responses including errors.<br>
```
"Text" not "test"

thc202 replied:
There's a typo, "text responses".

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/CookieHttpOnlyScannerUnitTest.java:
```diff
@@ -0,0 +1,154 @@
+				"Set-Cookie: test=123; Path=/;\r\n" +
+				"Content-Type: text/html;charset=ISO-8859-1\r\n" +
+				"Content-Length: " + msg.getResponseBody().length() + "\r\n");
+		String response = msg.getResponseHeader().toString() + msg.getResponseBody().toString();
```
I should have picked this up on yesterday's PR but this standard response setup should be extracted to a helper method, instead of repeating.

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/CookieHttpOnlyScannerUnitTest.java:
```diff
@@ -0,0 +1,154 @@
+				"Server: Apache-Coyote/1.1\r\n" +
+				"Set-Cookie: test=123; Path=/;\r\n" +
+				"Content-Type: text/html;charset=ISO-8859-1\r\n" +
+				"Content-Length: " + msg.getResponseBody().length() + "\r\n");
```
I should have picked this up on yesterday's PR but this standard response setup should be extracted to a helper method, instead of repeating.

psiinon replied:
You mean like SomeHelper.stdResponse("Set-Cookie: test=123; Path=/;"); ?
I kind of like seeing the whole header - I think its more useful. But happy to be outvoted ;)

kingthorin replied:
No I mean the chunk that's usually the same:
* status line
* server, though perhaps drop that?
* content type
* content length

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/CookieHttpOnlyScannerUnitTest.java:
```diff
@@ -0,0 +1,154 @@
+			}
+		};
+		rule = new CookieHttpOnlyScanner();
+		rule.setParent(parent );
```
Extra space could be removed

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/CookieHttpOnlyScannerUnitTest.java:
```diff
@@ -0,0 +1,154 @@
+	}
+
+    @Test
+	public void noHttpOnlyAttribute () throws HttpMalformedHeaderException {
```
Space between function name and brackets could be removed (throughout).

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/CookieHttpOnlyScannerUnitTest.java:
```diff
@@ -0,0 +1,154 @@
+		msg.setResponseHeader(
+				"HTTP/1.1 200 OK\r\n" +
+				"Server: Apache-Coyote/1.1\r\n" +
+				"Set-Cookie: hasatt=test123; Path=/; HttpOnly\r\n" +
```
Mark this one secure as well just to ensure parsing occurs correctly?

psiinon replied:
I'm not sure I follow - how is the secure flag relevant to this test?

kingthorin replied:
It would just ensure that the cookie is being parsed correctly when there's another attribute.

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/HeaderXssProtectionScannerUnitTest.java:
```diff
@@ -0,0 +1,262 @@
+    }
+
+    @Test
+	public void noXssProtectionCssLowNoBody () throws HttpMalformedHeaderException {
```
Where does the Css part of the function name come into play?

psiinon replied:
Ah, yeah, was playing around with different options ;)

thc202 commented on test/org/zaproxy/zap/extension/pscanrules/XFrameOptionsScannerUnitTest.java:
```diff
@@ -0,0 +1,333 @@
+				"HTTP/1.1 200 OK\r\n" +
+				"Server: Apache-Coyote/1.1\r\n" +
+				"X-Frame-Options: DENY\r\n" +
+				"X-Frame-Options: SAMEORIGIN\r\n" +
```
One of the headers should be removed.

thc202 replied:
Ignore, comment in wrong test.

thc202 replied:
One of the headers should be removed (the case of multiple headers is already tested).

thc202 replied:
The case of the meta tag and header is also being tested, it if this one is to really test the multiple headers and meta tag the name of the test should be changed (and the previous one too).

thc202 commented on test/org/zaproxy/zap/extension/pscanrules/CookieHttpOnlyScannerUnitTest.java:
```diff
@@ -0,0 +1,154 @@
+				"Content-Length: " + msg.getResponseBody().length() + "\r\n");
+		String response = msg.getResponseHeader().toString() + msg.getResponseBody().toString();
+		Source source = new Source(response);
+		rule.scanHttpResponseReceive(msg, rule.getPluginId(), source);
```
The seconds parameter is the ID of the history reference being scanned, probably better to use other number than the plugin ID for that.

thc202 commented on test/org/zaproxy/zap/extension/pscanrules/CookieHttpOnlyScannerUnitTest.java:
```diff
@@ -0,0 +1,154 @@
+	private PassiveScanThread parent;
+	private List<Alert> alertsRaised;
+
+    @BeforeClass
```
There are some places that's using spaces for indentation that should be changed to match the same chars as the rest of the code (same in other test classes).

psiinon commented:
Have addressed most of these comments, with the exception of a header helper method - about to post to the dev group thread about this.
Have also included fixes for zaproxy/zaproxy/#2732

kingthorin commented on src/org/zaproxy/zap/extension/pscanrules/CacheControlScanner.java:
```diff
@@ -50,7 +52,32 @@ public void scanHttpRequestSend(HttpMessage msg, int id) {
+					return;
+				}
+				if (msg.getRequestHeader().getURI().toString().toLowerCase().endsWith(".css")) {
+					
```
Should there be an action inside this if?

psiinon commented:
Helper method thread: https://groups.google.com/d/msg/zaproxy-develop/1m-Bznmg1uI/LHollYQ-BwAJ
CacheControlScanner thread: https://groups.google.com/d/msg/zaproxy-develop/6IAnEHANWpI/4xMfTkk-BwAJ

kingthorin commented on src/org/zaproxy/zap/extension/pscanrules/SetCookieUtils.java:
```diff
@@ -62,6 +65,85 @@ public static boolean hasAttribute(String headerValue, String attributeName) {
+    }
+
+    /**
+     * Returns the relevant SetCookie(2) line as well as just the cookie name, or null if not found.
```
What's the significance of `(2)`? Is that meant to indicate that this handles both Set-Cookie and Set-Cookie2? If so I'd suggest being specific here:

> Returns the relevant Set-Cookie or Set-Cookie2 line as well as...

The return statement should mention something like the SetCookie header plus name. Which should be clear enough once the description of the function is updated....my 2cents.

Also you might want to say header vs line, as line would seem to imply the whole header up to some sort of line ending or wordwrap.

kingthorin commented on src/org/zaproxy/zap/extension/pscanrules/resources/help/contents/pscanrules.html:
```diff
@@ -14,7 +14,9 @@
 <H2>Cache Control</H2>
-Checks "Cache-Control" and "Pragma" response headers against general industry best practice settings for protection of sensitive content. 
+Checks "Cache-Control" and "Pragma" response headers against general industry best practice settings for protection of sensitive content.<br>
+At MEDIUM and HIGH thresholds only non error or redirect text responses (excluding JavaScript) are considered.<br>
```
I suggest hyphenating `non-error`.

kingthorin commented on src/org/zaproxy/zap/extension/pscanrules/resources/help/contents/pscanrules.html:
```diff
@@ -29,13 +31,18 @@
 <H2>Header XSS Protection</H2>
 Checks for the existence of and value/setting of the X-XSS-Protection header. 
-This response header can be used to configure a user-agent's built in reflective XSS protection.
+This response header can be used to configure a user-agent's built in reflective XSS protection.<br>
```
This is a pre-existing error but shouldn't "built in" be one word or hyphenated?

kingthorin commented on src/org/zaproxy/zap/extension/pscanrules/resources/help/contents/pscanrules.html:
```diff
@@ -29,13 +31,18 @@
 Checks for the existence of and value/setting of the X-XSS-Protection header. 
-This response header can be used to configure a user-agent's built in reflective XSS protection.
+This response header can be used to configure a user-agent's built in reflective XSS protection.<br>
+At MEDIUM and HIGH thresholds only non error or redirect HTML responses are considered.<br>
```
I suggest hyphenating `non-error`.

I think what you want here is:

> At MEDIUM and HIGH thresholds only non-error or non-redirect HTML...

kingthorin commented on src/org/zaproxy/zap/extension/pscanrules/resources/help/contents/pscanrules.html:
```diff
@@ -55,14 +62,16 @@
-does not alert on client or server error responses, for all other enabled thresholds it alerts on all response 
-types. The following conditions may result in an alert:
+This scanner checks for the existence and validity of the X-Frame-Options header.<br> 
+At MEDIUM and HIGH thresholds this only looks at non error or redirect HTML responses.<br>
```
I suggest hyphenating `non-error`.

I think what you want here is:

> At MEDIUM and HIGH thresholds only non-error or non-redirect HTML...

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/CacheControlScannerUnitTest.java:
```diff
@@ -0,0 +1,348 @@
+        
+        msg.setResponseBody("<html></html>");
+        msg.setResponseHeader(
+                "HTTP/1.1 301 OK\r\n" +
```
Is the status text at all relevant? By spec/convention shouldn't this actually be `HTTP/1.1 301 Moved Permanently`?

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/CacheControlScannerUnitTest.java:
```diff
@@ -0,0 +1,348 @@
+        
+        msg.setResponseBody("<html></html>");
+        msg.setResponseHeader(
+                "HTTP/1.1 301 OK\r\n" +
```
Same as earlier

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/CacheControlScannerUnitTest.java:
```diff
@@ -0,0 +1,348 @@
+        rule.setLevel(AlertThreshold.MEDIUM);
+
+        HttpMessage msg = new HttpMessage();
+        msg.setRequestHeader("GET https://www.example.com/test/ HTTP/1.1");
```
Function name implies JS test.

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/CacheControlScannerUnitTest.java:
```diff
@@ -0,0 +1,348 @@
+        HttpMessage msg = new HttpMessage();
+        msg.setRequestHeader("GET https://www.example.com/test/ HTTP/1.1");
+        
+        msg.setResponseBody("<html></html>");
```
Function name implies JS test.

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/CacheControlScannerUnitTest.java:
```diff
@@ -0,0 +1,348 @@
+        rule.setLevel(AlertThreshold.HIGH);
+
+        HttpMessage msg = new HttpMessage();
+        msg.setRequestHeader("GET https://www.example.com/test/ HTTP/1.1");
```
Function name implies JS test.

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/CacheControlScannerUnitTest.java:
```diff
@@ -0,0 +1,348 @@
+        HttpMessage msg = new HttpMessage();
+        msg.setRequestHeader("GET https://www.example.com/test/ HTTP/1.1");
+        
+        msg.setResponseBody("<html></html>");
```
Function name implies JS test.

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/CookieHttpOnlyScannerUnitTest.java:
```diff
@@ -0,0 +1,127 @@
+        assertThat(alertsRaised.get(0).getParam(), equalTo("test"));
+        assertThat(alertsRaised.get(0).getEvidence(), equalTo("Set-Cookie: test"));
+    }
+    
```
Should we have a test cases that sets multiple cookies in a single Set-Cookie header?

Something like:
`Set-Cookie: test=b;Path=/;,test2=d;Path=/;HttpOnly;Secure`

psiinon replied:
Yes, and it failed (reported wrong cookie) - so good suggestion!

psiinon replied:
OK, supporting this looks like it could be a nightmare to parse correctly.
And the latest RFC https://tools.ietf.org/html/rfc6265#section-3 states that "Origin servers SHOULD NOT fold multiple Set-Cookie header fields into a single header field." On this basis I'm proposing that we dont support it, at least not at this stage. This is not a regression, we didnt support it before;)

kingthorin replied:
Ok, that seems fair.

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/CookieSecureFlagScannerUnitTest.java:
```diff
@@ -0,0 +1,128 @@
+        assertThat(alertsRaised.get(0).getParam(), equalTo("test"));
+        assertThat(alertsRaised.get(0).getEvidence(), equalTo("Set-Cookie: test"));
+    }
+    
```
Should we have a test case that sets multiple cookies in a single Set-Cookie header?

Something like:
`Set-Cookie: test=b;Path=/;,test2=d;Path=/;HttpOnly;Secure`

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/CrossDomainScriptInclusionScannerUnitTest.java:
```diff
@@ -0,0 +1,349 @@
+
+    @Test
+    public void crossDomainScriptWithIntegrity () throws HttpMalformedHeaderException {
+        // Mock the model and session
```
Do Model and Session matter when Integrity is in play? I didn't think they did.

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/HeaderXssProtectionScannerUnitTest.java:
```diff
@@ -0,0 +1,306 @@
+	}
+
+    @Test
+    public void noXssProtection () throws HttpMalformedHeaderException {
```
Seems to be an extra space between the function name and brackets, throughout.

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/XFrameOptionsScannerUnitTest.java:
```diff
@@ -0,0 +1,392 @@
+	}
+
+    @Test
+    public void xframeOptionsDeny () throws HttpMalformedHeaderException {
```
Seems to be an extra space here, and throughout this test set.

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/XFrameOptionsScannerUnitTest.java:
```diff
@@ -0,0 +1,392 @@
+        
+        msg.setResponseBody("<html></html>");
+        msg.setResponseHeader(
+                "HTTP/1.1 401 OK\r\n" +
```
401?

kingthorin replied:
`401 Unauthorized`

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/XFrameOptionsScannerUnitTest.java:
```diff
@@ -0,0 +1,392 @@
+        HttpMessage msg = new HttpMessage();
+        msg.setRequestHeader("GET http://www.example.com/test/ HTTP/1.1");
+        
+        msg.setResponseBody("<html></html>");
```
Function name and content-type suggest text...while I suppose this isn't wrong (the text could be html tags), it's kind of misleading.

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/XFrameOptionsScannerUnitTest.java:
```diff
@@ -0,0 +1,392 @@
+        HttpMessage msg = new HttpMessage();
+        msg.setRequestHeader("GET http://www.example.com/test/ HTTP/1.1");
+        
+        msg.setResponseBody("<html></html>");
```
Function name and content-type suggest text...while I suppose this isn't wrong (the text could be html tags), it's kind of misleading.

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/XFrameOptionsScannerUnitTest.java:
```diff
@@ -0,0 +1,392 @@
+        HttpMessage msg = new HttpMessage();
+        msg.setRequestHeader("GET http://www.example.com/test/ HTTP/1.1");
+        
+        msg.setResponseBody("<html></html>");
```
Function name and content-type suggest text...while I suppose this isn't wrong (the text could be html tags), it's kind of misleading.

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/XFrameOptionsScannerUnitTest.java:
```diff
@@ -0,0 +1,392 @@
+        msg.setResponseHeader(
+                "HTTP/1.1 200 OK\r\n" +
+                "Server: Apache-Coyote/1.1\r\n" +
+                "X-Frame-Options: DENY\r\n" +
```
This line should be dropped, the next test covers meta+header.

psiinon commented:
Hopefully fixed all the comments.
Also changed to only report X-Frame-Options issues at LOW threshold if CSP 'frame-ancestors' element present.

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/XContentTypeOptionScannerUnitTest.java:
```diff
@@ -0,0 +1,175 @@
+        
+        msg.setResponseBody("<html></html>");
+        msg.setResponseHeader(
+                "HTTP/1.1 301 OK\r\n" +
```
One seems to have squeaked by `Moved Permanently`

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/CacheControlScannerUnitTest.java:
```diff
@@ -0,0 +1,348 @@
+        HttpMessage msg = new HttpMessage();
+        msg.setRequestHeader("GET https://www.example.com/test.js HTTP/1.1");
+        
+        msg.setResponseBody("<html></html>");
```
This could be removed, or a dummy string used (JS test).

kingthorin commented on test/org/zaproxy/zap/extension/pscanrules/CacheControlScannerUnitTest.java:
```diff
@@ -0,0 +1,348 @@
+        HttpMessage msg = new HttpMessage();
+        msg.setRequestHeader("GET https://www.example.com/test.js HTTP/1.1");
+        
+        msg.setResponseBody("<html></html>");
```
This could be removed, or a dummy string used (JS test).

kingthorin commented:
The CSP thing looks good, made a few other small notes, there are still a few places that there's a space between function name and brackets. Looking good.

Thanks for tackling this @psiinon.

psiinon commented:
Hopefully fixed those comments

psiinon commented:
Updated to reduce the boilerplate code in the unit tests c/o @thc202

psiinon commented:
More tweaks based on feedback

thc202 commented:
Looks good to me.
